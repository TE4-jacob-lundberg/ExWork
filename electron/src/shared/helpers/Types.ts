export interface ILink {
  url: string;
  label: string;
}

export interface IGameFormFields {
  id: string;
  title: string;
  abbreviation: string;
  image: string; 
  xAxis: string;
  yAxis: string;
  links: ILink[];
  fileNames: string[];
}
