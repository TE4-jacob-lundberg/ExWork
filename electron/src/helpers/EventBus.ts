/* eslint-disable @typescript-eslint/no-explicit-any */
const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

export { EventBus };
