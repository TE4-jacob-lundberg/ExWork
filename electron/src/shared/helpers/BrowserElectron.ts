// eslint-disable-next-line @typescript-eslint/no-explicit-any
export let electron: any;

if (process.env.ENVIRONMENT === 'test') electron = require('electron');
else electron = window.require('electron');
