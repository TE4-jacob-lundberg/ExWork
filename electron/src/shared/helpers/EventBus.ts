/* eslint-disable @typescript-eslint/no-explicit-any */
interface IEventBus {
  events: {[id: string]: Function[]};
  dispatch: (event: string, data: any) => void;
  subscribe: (event: string, callBack: Function) => void;
}

const EventBus: IEventBus = {
  events: {},
  dispatch: function (event: string, data: any): void {
    if (!this.events[event]) return;
    this.events[event].forEach((callBack: Function) => callBack(data));
  },
  subscribe: function (event: string, callBack: Function): void {
    if (!this.events[event]) this.events[event] = [];
    this.events[event].push(callBack);
  },
};

export { EventBus };
