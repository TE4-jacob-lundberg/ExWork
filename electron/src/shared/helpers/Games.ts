import { standardGames, IGame } from '../constants/standardGames';

export class Games {

  public static all(): IGame[] {
    if (!localStorage.getItem('games')) localStorage.setItem('games', JSON.stringify(standardGames));

    return JSON.parse(localStorage.getItem('games')!) || [];
  }

  public static addNew(game: IGame): string {
    const savedGames: IGame[] = JSON.parse(localStorage.getItem('games')!);
    if (savedGames.some(sGame => sGame.id === game.id)) return 'DUPLICATE_ID';
    
    localStorage.setItem('games', JSON.stringify(savedGames.concat(game)));
    return '';
  }
}
