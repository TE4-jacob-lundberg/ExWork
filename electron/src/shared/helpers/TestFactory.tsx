/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createBrowserHistory } from 'history';

export function factory(component: JSX.Element): RenderResult {
  return render(component);
}

export function routerFactory(component: JSX.Element, options: any = { history: createBrowserHistory() }): RenderResult {
  return factory(
    <Router history={options.history}>
      {component}
    </Router>
  );
}
