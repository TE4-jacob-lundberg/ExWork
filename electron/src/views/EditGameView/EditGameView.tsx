import React, { useEffect, useState } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { useIndexedDB } from 'react-indexed-db';

import GameFormComponent from '../../shared/components/GameFormComponent';
import PageTitleComponent from '../../shared/components/PageTitleComponent';
import ModalComponent from '../../shared/components/ModalComponent';
import { routes } from '../../shared/constants/routes';
import { IGame } from '../../shared/helpers/Types';

interface Props {}

interface Params {
  gameID: string;
}

const AddGameView: React.FC<Props> = function () {
  const params = useParams<Params>();
  const [initialGame, setInitialGame] = useState<IGame>();
  const [showModal, setShowModal] = useState(false);
  const history = useHistory();
  const db = useIndexedDB('games');

  useEffect(() => {
    if (initialGame) return;
    db.getByID(params.gameID).then((resp: IGame) => setInitialGame(resp));
  }, [db, initialGame, params.gameID]);

  return (
    <> 
      <PageTitleComponent title="Edit game" />
      {initialGame && <GameFormComponent 
        initialValues={{
          title: initialGame.title,
          abbreviation: initialGame.abbreviation || '',
          image: initialGame.image.url,
          xAxis: initialGame.image.bannerPos.x,
          yAxis: initialGame.image.bannerPos.y,
          fileNames: initialGame.fileNames,
        }}
        onSubmit={(values): void => {
          db.update({
            ...values, 
            id: initialGame.id,
            links: initialGame.links,
            image: {
              url: values.image, 
              bannerPos: { 
                x: values.xAxis.toString(), 
                y: values.yAxis.toString(),
              },
            },
          }).then(() => history.push(routes.showGame.replace(':gameID', initialGame.id)));
        }}
        onDelete={(): void => {
          setShowModal(true);
        }}
        deletable
      />}
      {showModal && initialGame && <ModalComponent 
        title="Deleting a game"
        onClose={(): void => setShowModal(false)}
        continueLabel="Yes, delete it"
        onContinue={(): void => {
          db.deleteRecord(initialGame.id).then(() => history.push(routes.games));
        }}
        cancelLabel="No, go back"
        onCancel={(): void => setShowModal(false)}
        cancelAble
      >
        <h3>{`Are you sure you want to delete "${initialGame!.title}"`}</h3>  
      </ModalComponent>}
    </>
  );
};

export default AddGameView;
