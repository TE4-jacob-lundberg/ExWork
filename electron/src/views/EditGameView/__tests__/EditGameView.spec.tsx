import React from 'react';
import { fireEvent, waitForElement } from '@testing-library/react';
import { createMemoryHistory } from 'history';

import { routerFactory } from '../../../shared/helpers/TestFactory';
import { IGame } from '../../../shared/helpers/Types';
import EditGameView from '../EditGameView';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: (): object => ({
    gameID: '0',
  }),
}));

jest.mock('react-indexed-db', () => ({
  useIndexedDB: (): object => {
    return {
      getByID: (): Promise<IGame> => Promise.resolve({
        id: '1',
        title: 'testTitle',
        abbreviation: 'testAbbr',
        image: {
          url: 'test',
          bannerPos: {
            x: '0',
            y: '0',
          },
        },
        links: [
          { type: 'link', label: 'testWiki', url: 'test.com'},
        ],
        fileNames: ['test.exe'],
      }),
      deleteRecord: (): Promise<void> => Promise.resolve(),
    };
  },
}));

const testComponent = <EditGameView />;

describe('EditGameView.tsx', () => {
  it('renders correctly', async () => {
    const { container, getByText } = routerFactory(testComponent);

    await waitForElement(() => getByText('Save'));
    expect(container).toMatchSnapshot();
  });

  it('allows for game to be deleted', async () => {
    const history = createMemoryHistory();
    const { getByText, queryByText, getByTestId } = routerFactory(testComponent, {history});

    await waitForElement(() => getByText('Save'));
    fireEvent.click(getByText('Delete'));
    expect(getByText('Deleting a game')).toBeTruthy();

    fireEvent.click(getByTestId('continue-modal'));

    await waitForElement(() => queryByText('Deleting a game'));
    expect(history.location.pathname).toBe('/games');
  });
});
