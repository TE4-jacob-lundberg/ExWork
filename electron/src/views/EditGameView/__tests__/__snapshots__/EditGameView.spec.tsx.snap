// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditGameView.tsx renders correctly 1`] = `
<div>
  <div
    class="css-vnxadl"
    data-testid="page-title"
  >
    <h1
      class="css-g7rarf"
    >
      Edit game
    </h1>
  </div>
  <div
    class="css-o6jotw"
  >
    <form>
      <h1
        class="css-1vjmso2"
      >
        Game Information
      </h1>
      <section
        class="css-6a1pbd"
      >
        <label
          class="css-13m3dri"
          for="title"
        >
          Title*
        </label>
        <input
          class="css-1o0il4p"
          id="title"
          name="title"
          type="text"
          value="testTitle"
        />
        <label
          class="css-13m3dri"
          for="abbr"
        >
          Abbreviation
        </label>
        <input
          class="css-1o0il4p"
          id="abbr"
          name="abbreviation"
          type="text"
          value="testAbbr"
        />
        <label
          class="css-13m3dri"
        >
          Game files (*.exe, *.app)
        </label>
        <input
          class="css-1o0il4p"
          name="fileNames"
          type="text"
          value="test.exe"
        />
      </section>
      <h1
        class="css-1vjmso2"
      >
        Image Settings
      </h1>
      <section
        class="css-6a1pbd"
      >
        <label
          class="css-13m3dri"
        >
          Image*
        </label>
        <input
          class="css-1o0il4p"
          type="file"
        />
        <label
          class="css-13m3dri"
        >
          X-axis
        </label>
        <input
          class="css-jq1eod"
          max="100"
          min="0"
          name="xAxis"
          type="range"
          value="0"
        />
        <label
          class="css-13m3dri"
        >
          Y-axis
        </label>
        <input
          class="css-jq1eod"
          max="100"
          min="0"
          name="yAxis"
          type="range"
          value="0"
        />
      </section>
      <div
        class="css-468wco"
      >
        <button
          class="css-34a7o7"
        >
          Delete
          <i
            class="material-icons"
          >
            delete
          </i>
        </button>
        <button
          class="css-1ctwf7v"
          data-testid="submit-form"
          type="submit"
        >
          Save
          <i
            class="material-icons"
          >
            save
          </i>
        </button>
      </div>
    </form>
    <div
      class="css-npfm1w"
      x="0"
      y="0"
    />
  </div>
</div>
`;
